pkg/gce-pd-csi-driver/controller.go

```
@@ -599,11 +609,23 @@ func (gceCS *GCEControllerServer) createSingleDeviceDisk(ctx context.Context, re
 func (gceCS *GCEControllerServer) createSingleDisk(ctx context.Context, req *csi.CreateVolumeRequest, params common.DiskParameters, volKey *meta.Key, zones []string) (*gce.CloudDisk, error) {
 	capacityRange := req.GetCapacityRange()
 	capBytes, _ := getRequestCapacity(capacityRange)
-	multiWriter, _ := getMultiWriterFromCapabilities(req.GetVolumeCapabilities())
 	readonly, _ := getReadOnlyFromCapabilities(req.GetVolumeCapabilities())
-	accessMode := params.AccessMode
+	accessMode := ""
+	multiWriter := false
+	if common.IsHyperdisk(params.DiskType) {
+		if !disksWithUnsettableAccessMode[params.DiskType] {
+			if am, err := getHyperdiskAccessModeFromCapabilities(req.GetVolumeCapabilities()); err != nil {
```

Just making sure I understand, this check is actually redundant, as we should prevent unsettable access modes from getting set up in createVolumeInternal?

---

We need this because we could still derive "READ_WRITE_SINGLE" from the volume capabilities, which is a valid mode for these disks; createVolumeInternal only throws error if the volume capabilities contain multi-node access modes (namely "READ_WRITE_MANY" and "READ_ONLY_MANY").

---

So if this is an unsettable type (HdT etc), then we'll silently ignore req.GetVolumeCapabilities (or at least the access mode from them)? I'm not sure that's right, I think we should error if disksWithUnsettableAccessMode && am != RW_SINGLE.

Or am I getting confused about something?

---

That check is performed somewhere else (see other comment). Let me know if you think there's improvement I can make here to make the code more understandable.

---

I think we should return an error if disksWithUnsettableAccessMode[params.DiskType] == true here, even if it's redundant with the check above. It will save us if there is a future refactor that changes the order of validation.

---

Good point, done.